### DISCUSSION TITLE
What is the exact danger of not waiting for peer's close_notify response?

### DISCUSSION THREAD

**Discussion by dragan-ponos** [openssl/openssl#23630] [Q&A] [answered]
Hello,

From the documentation:

> Alternatively, it is acceptable for an application to call SSL_shutdown() once (such that it returns 0) and then close the underlying connection without waiting for the peer's response. This allows for a more rapid shutdown process if the application does not wish to wait for the peer.

> **This alternative "fast shutdown" approach should only be done if it is known that the peer will not send more data, otherwise there is a risk of an application exposing itself to a truncation attack.**

Could you please elaborate on the dangers?
In my case, I am handling the case where a higher layer is closing the connection with no intent to read any more data.

Maybe the higher layer considers this data to be discardable, or maybe it has a bug in it. Regardless of the reason, from the perspective of the TLS layer, I think "fast shutdown" would be perfectly fine. What am I missing?

Thanks!

**Comment by t8m** [comment, marked as answer]
If there is no intent to read any further data from the peer, then it is of course safe to not wait for the close_notify, unless you want to detect a possible situation where the peer sent some bogus additional data to you. If you do not care detecting such misbehaving peer, there is no need to wait.

**Comment by kroeckx** [comment]
Examples where there is no problem:
- You know it's going to send certain amount of bytes, and you've received all those bytes, for instance because it first send the length.
- There is some way that you can tell all data has been received, for instance it's an xml file, and you've found the closing </xml>.

It all really depends on the protocol and the data involved, and how strict the implementation is.

If you have a combination of protocol and data where you can't tell all data was received, TLS allows you to make sure that everything was received.

In general, it's better to make sure that at the protocol it's clear that everything was received. For the case where the sender wants to know the receiver has received it, it's really best to do this at the protocol level.

**Reply by dragan-ponos** [reply]
I think that both documentation and you, @kroeckx, talk about closing connection from below and passing EOF up to the application layer.

However, in my scenario, my higher layer wants to close the connection (both directions) with no further processing, and I want **to write my TLS layer that will cover this securely**. Would you agree that this would be the third example where there is no problem, which is what I got from @t8m?

If I am correct, I would reword the documentation as:
> This alternative "fast shutdown" approach should only be done if it is known that the peer will not send more data, **or the application has no interest in reading any more data**, otherwise there is a risk of an application exposing itself to a truncation attack.

Thank you both for explanation.

**Reply by kroeckx** [reply]
If both sides agree that no more data will be sent, both sides can just close the TCP connection without sending the close_notify, or waiting for it. The close_notify just provides an extra layer that says all data has been sent, in case you can't do it at the application layer.

So yes, if the application doesn't care about any extra data the sender might sent, it can also just close the connection.

**Comment by dragan-ponos** [comment]
Thank you for your time. I was reading more on the topic. I think the documentation was right and I was wrong. While dropping data would not create problems in my code, TLS control messages, such as new session ticket in TLS 1.3, would also be dropped by a "fast shutdown". Better to do a proper shutdown.
